### PR TITLE
Homework 6

### PR DESCRIPTION
**Pull request by shivakcpm**
* Add Redux Actions
* Add rest calls
* Add async actions

### CODE REVIEW

File: src/components/ContentHolder/ContentHolderComponent.js [hunk 1]
```diff
@@ -1,71 +1,67 @@
...
+import React, {  useState, useCallback } from 'react';
 import TabComponent from '../TabComponent/TabComponent';
 import MovieListComponent from '../MovieList/MovieListComponent';
 import { ErrorBoundary } from '../ErrorBoundary/ErrorBoundary';
-import { Context } from '../../contextProvider';
+import { HeaderComponent } from '../HeaderComponent/HeaderComponent';
+import MovieDetails from '../MovieDetails/MovieDetailsComponent';
 import { TABS, DEFAULT_SORT_BY, DEFAULT_FILTER_BY } from '../../consts/constants';
+import { store, getMoviesActionAsync } from '../../store/store';
+import { connect } from 'react-redux';
```
Comment (alionaSipchuk) [inline src/components/ContentHolder/ContentHolderComponent.js:9]: let's add import from libraries to the top of imports list)
move it after import react))

File: src/components/ContentHolder/ContentHolderComponent.js [hunk 2]
```diff
@@ -1,71 +1,67 @@
...
+  const [movieDetails, setMovieDetails] = useState(null);
+  const [query, setQuery] = useState('');
 
   const onTabChange = index => {
     setGenre(TABS[index]);
   };
 
+  const fetchMovies = () => {
+    const filterBy = genre !== DEFAULT_FILTER_BY
+      ? [genre]
```
Comment (alionaSipchuk) [inline src/components/ContentHolder/ContentHolderComponent.js:24]: possible no need to wrap this value in the array, because in one time we have only one filtering value

File: src/components/ContentHolder/ContentHolderComponent.js [hunk 3]
```diff
@@ -1,71 +1,67 @@
...
+      : null;
+    store.dispatch(getMoviesActionAsync({search:query, searchBy:'title', sortOrder:'desc', sortBy,
```
Comment (alionaSipchuk) [inline src/components/ContentHolder/ContentHolderComponent.js:26]: move each props into the next line, it's will be better to read)

File: src/components/ContentHolder/ContentHolderComponent.js [hunk 4]
```diff
@@ -1,71 +1,67 @@
...
+  }, [genre, sortBy, query]);
 
   const handleSort = event => {
     const sortedBy = event.target.value === DEFAULT_SORT_BY
       ? DEFAULT_SORT_BY
       : 'title';
-    setSortOrder(sortedBy);
+      setSortBy(sortedBy);
   };
+genreChange();
```
Comment (alionaSipchuk) [inline src/components/ContentHolder/ContentHolderComponent.js:40]: why did you call here?? each time that it will call all items of component will be rerenders, no need to do it, we can avoid it!! 

because we try to avoid this by adding HOOKS!))

File: src/components/HeaderComponent/HeaderComponent.js [hunk 5]
```diff
@@ -16,7 +17,8 @@ export class HeaderComponent extends PureComponent {
 
   addMovie = movie => {
     this.toggleModel();
-    this.props.setAllMovies([...this.props.allMovies, movie]);
+    store.dispatch(addMovieActionAsync(movie));
```
Comment (alionaSipchuk) [inline src/components/HeaderComponent/HeaderComponent.js:20]: i think no need to add word action for the name of action, pls move it)

File: src/components/MovieCard/MovieCardComponent.js [hunk 6]
```diff
@@ -46,20 +46,20 @@ export default class MovieCardComponent extends PureComponent {
   };
 
   onDelete = () => {
-    this.props.deleteMovie(this.props.movie);
+    this.props.deleteMovie({id:this.props.movie.id});
```
Comment (alionaSipchuk) [inline src/components/MovieCard/MovieCardComponent.js:49]: we can make it more simple:

```
const { movie: { id } } = this.props;
this.props.deleteMovie({ id });
```

File: src/components/MovieCard/MovieCardComponent.js [hunk 7]
```diff
@@ -69,9 +69,9 @@ export default class MovieCardComponent extends PureComponent {
         <div className="card-footer">
           <div className="title-info">
             <div className="title">{title}</div>
-            <div className="year">{new Date(releaseDate).getFullYear()}</div>
+            <div className="year">{new Date(release_date).getFullYear()}</div>
           </div>
-          <div className="category">{category}</div>
+          <div className="category">{genres.join(', ')}</div>
```
Comment (alionaSipchuk) [inline src/components/MovieCard/MovieCardComponent.js:74]: i recommend you to move it into separate const and wrap into the useMemo hook!)
```
const generesList = useMemo(
  () => genres.join(', '),
  [genres],
);
```

File: src/components/MovieDetails/MovieDetailsComponent.js [hunk 8]
```diff
@@ -14,12 +14,12 @@ const MovieDetails = (props) => {
...
         <div className="movie-details-parent">
-          <img className="details-image" src={props.src} />
+          <img className="details-image" src={props.poster_path} />
           <div className="details-wrapper">
             <span className="movie-title">{props.title}</span>
-            <span className="rating">{props.rating}</span>
-            <div className="category">{props.category}</div>
-            <span className="release-date color-red">{new Date(props.releaseDate).getFullYear()}</span>
+            <span className="rating">{props.vote_average}</span>
+            <div className="category">{props.genres.join(', ')}</div>
```
Comment (alionaSipchuk) [inline src/components/MovieDetails/MovieDetailsComponent.js:21]: as you call join with same separator, you can move it into one formatter and use it averywhere it need))

File: src/services/movieServices.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        dispatch(actions.getMoviesAction({ movies:data }));
+    }, () => {
+        dispatch(actions.getMoviesAction({ movies:{} }));
+    });
+}
+
+
+export function editMovie(dispatch, payload) {
+    const url = URLS.getUrl(URLS.MOVIES_URL, null, null);
+    return fetch(url, {method:'PUT', headers:{'Content-Type': 'application/json'}, body:JSON.stringify(payload)}).then(data => {
```
Comment (alionaSipchuk) [inline src/services/movieServices.js:18]: pls transfer each **then** to the new line, for best reading code)
in the whole file)

File: src/services/movieServices.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    }).then((data) => {
+        dispatch(actions.editMovieAction({ movies:data }));
+    }, () => {
+        dispatch(actions.editMovieAction({ movies:{} }));
+    });
+}
+
+export function addMovie(dispatch, payload) {
+    const url = URLS.getUrl(URLS.MOVIES_URL, null, null);
+    return fetch(url, {method:'POST', headers:{'Content-Type': 'application/json'}, body:JSON.stringify(payload)}).then(data => {
```
Comment (alionaSipchuk) [inline src/services/movieServices.js:29]: let's create **stringifyValue** function, and use it for each transform body case)

File: src/services/movieServices.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+export function deleteMovie(dispatch, pathParams) {
+    const url = URLS.getUrl(URLS.MOVIE_BY_ID, pathParams, null);
+    return fetch(url, {
+        method: 'delete'
+    }).then(() => {
+        dispatch(actions.deleteMovieAction({movies:{} }));
+    }, () => {
+        dispatch(actions.deleteMovieAction({movies:{} }));
+    });
```
Comment (alionaSipchuk) [inline src/services/movieServices.js:47]: it's duplicate)

File: src/store/rootReducer.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+
```
Comment (alionaSipchuk) [inline src/store/rootReducer.js:1]: remove empty line

File: src/components/MovieList/MovieListComponent.js [hunk 13]
```diff
@@ -1,13 +1,24 @@
...
-  }
+
+  const editMovie = movie => {
+    store.dispatch(editMovieActionAsync(movie)).then(() => {
+      props.fetchMovies();
+    });
+  };
+
+  const deleteMovie =  movie => {
+    store.dispatch(deleteMovieActionAsync(movie)).then(() => {
```
Comment (alionaSipchuk) [inline src/components/MovieList/MovieListComponent.js:18]: what i want to recommend you, let's try to move  store.dispatch method into EACH action inside))

and let's use _async and await_ for promise)

File: src/store/rootReducer.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+import * as actions from './actions';
+
+const initialState = {
+    movies:{}
```
Comment (alionaSipchuk) [inline src/store/rootReducer.js:5]: why did with default value object and not an array?

File: src/store/store.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
+import { createStore, applyMiddleware } from 'redux';
+import * as services from '../services/movieServices';
+import rootReducer from './rootReducer';
+import thunk from 'redux-thunk';
```
Comment (alionaSipchuk) [inline src/store/store.js:4]: also add import of thunk to the top of the list(after redux);

**Review by alionaSipchuk — CHANGES_REQUESTED** [review]
Not bad, has some not critical issues)

File: src/components/AddMovie/AddMovieComponent.js [hunk 16]
```diff
@@ -13,16 +13,27 @@ export default class AddMovieComponent extends Component {
     this.setState({
       movie: {
         ...this.state.movie,
-        [name]: value
+        [name]: name === 'runtime'
+          ? +value
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/AddMovie/AddMovieComponent.js:17]: we are getting number as string and api accepting only numbers.  using explicit conversion as suggested

File: src/components/AddMovie/AddMovieComponent.js [hunk 17]
```diff
@@ -13,16 +13,27 @@ export default class AddMovieComponent extends Component {
...
+          : value
       }
     });
   };
 
+  handleChange = ({ target: {selectedOptions, name } }) => {
+  this.setState({
+    movie: {
+      ...this.state.movie,
+      [name]: Array.from(selectedOptions, (item) => item.value)
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/AddMovie/AddMovieComponent.js:27]: selectedOptions is  oftype htmlcollection not array. one way is using foreach.

File: src/store/rootReducer.js (hunk 14 above)
Comment (shivakcpm) [reply src/store/rootReducer.js:5]: here movies as data and page object as two keys. name might be misleading i will rename it

File: src/components/AddMovie/AddMovieComponent.js [hunk 18]
```diff
@@ -67,20 +78,18 @@ export default class AddMovieComponent extends Component {
...
-            editMode
-              ? 'Edit Movie'
-              : 'Add Movie'
-          }
-        </div>
+       <div className="addmovie-header">
+        {editMode
+            ? 'Edit Movie'
+            : 'Add Movie'
+        }</div>
```
Comment (alionaSipchuk) [inline src/components/AddMovie/AddMovieComponent.js:92]: add tab! wrong formatting

File: src/components/AddMovie/AddMovieComponent.js [hunk 19]
```diff
@@ -196,11 +205,9 @@ export default class AddMovieComponent extends Component {
...
             }}
           >
-            {
-              editMode
-                ? 'SAVE'
-                : 'SUBMIT'
-            }
+            {editMode
+              ? 'SAVE'
+              : 'SUBMIT'}
```
Comment (alionaSipchuk) [inline src/components/AddMovie/AddMovieComponent.js:210]: pls revert this change!)

**Review by alionaSipchuk — APPROVED** [review]
